File: Leetcode/src/com/leetcode/math/_386_LexicographicalNumbers.java
package com.leetcode.math;

import java.util.ArrayList;
import java.util.List;

/**
 * _386_LexicographicalNumbers
 */
public class _386_LexicographicalNumbers {

    // backtracking。
    public List<Integer> lexicalOrder(int n) {
        List<Integer> res = new ArrayList<>();
        for (int i = 1; i < 10; i++) {
            dfs(res, n, i);
        }
        return res;
    } 
    
    public void dfs(List<Integer> res, int n, int cur) {
        if (cur > n) return;
        else {
            res.add(cur);
            for (int i = 0; i < 10; i++) {
                if (cur * 10 + i > n) break;
                dfs(res, n, cur * 10 + i);
            }
        }
    }
}

File: Leetcode/src/com/leetcode/math/_386_lexicographical_numbers.py
from typing import List


class Solution:
    """
    从1开始尝试，每次看当前数乘以10的情况，如果是大于n
    则代表我们需要移除最后一位，然后加上1
    """

    def lexicalOrder(self, n: int) -> List[int]:
        res = [0] * n
        curr = 1
        for i in range(n):
            res[i] = curr
            curr *= 10
            while curr > n:
                curr //= 10
                curr += 1
                while curr % 10 == 0:
                    curr //= 10
        return res

